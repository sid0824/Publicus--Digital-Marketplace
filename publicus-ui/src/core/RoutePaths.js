// The paths of every page in the site
export default {
  Home: '/',
  FormCreator: '/formCreator',
  CreateAccount: '/createAccount',
  Login: '/login',
  Marketplace : '/marketplace',
  MyForms: '/myForms',
  CreateForm: '/createForm',
  ViewResults: '/viewResults',
  CompleteForm: '/completeForm',
  MyAccount: '/myAccount',
  Administration: '/administration'
}