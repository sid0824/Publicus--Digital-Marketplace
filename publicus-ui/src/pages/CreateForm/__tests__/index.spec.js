import toJson from 'enzyme-to-json';
import React from "react";
import {shallow, configure} from 'enzyme';
import { message } from 'antd';
import Adapter from 'enzyme-adapter-react-16';
import {CreateForm, InputTypes} from "../index";
import {API} from "aws-amplify";
import Forms from '../../../fixtures/Forms';
import RoutePaths from "../../../core/RoutePaths";

configure({ adapter: new Adapter() });

const mockValues = {
  MOCK: 'VALUE',
};

const mockGetFieldDecorator = jest.fn();
const mockForm = {
  validateFields: jest.fn().mockImplementation(async (func) => func(null, mockValues)),
  getFieldDecorator: mockGetFieldDecorator,
};

const mockLocation = {
  search: "?surveyId=4fe4b2c0-f4a5-11e9-87b0-6785951dc893"
};

const mockPush = jest.fn();
const mockHistory = {
  push: mockPush
};

describe('Create Form', () => {

  beforeEach(() => {
    message.success = jest.fn();
    message.error = jest.fn();
    API.get = jest.fn().mockResolvedValue(Forms[2]);
    API.put = jest.fn().mockResolvedValue(null);
    API.post = jest.fn().mockResolvedValue(null);
    mockGetFieldDecorator.mockReturnValue((component) => component);
  });

  afterEach(() => {
    API.get.mockReset();
    API.put.mockReset();
    API.post.mockReset();
    message.error.mockReset();
    message.success.mockReset();
  });

  test('should render the create form page', (done) => {
    const wrapper = shallow(
      <CreateForm
        form={mockForm}
        location={mockLocation}
        history={mockHistory}
      />
    );
    setTimeout(() => {
      expect(toJson(wrapper)).toMatchSnapshot();
      done();
    });
  });

  test('should give error message on invalid form loaded', () => {
    API.get = jest.fn().mockImplementation(() => {
      throw new Error();
    });
    shallow(
      <CreateForm
        form={mockForm}
        location={mockLocation}
        history={mockHistory}
      />
    );
    expect(message.error).toHaveBeenCalledWith('Failed to load form data');
  });

  test('should generate initial configs for all input types', () => {
    const wrapper = shallow(
      <CreateForm
        form={mockForm}
        location={{
          search: ''
        }}
        history={mockHistory}
      />
    );
    wrapper.instance().handleAddNew({ key: InputTypes.TextInput });
    wrapper.instance().handleAddNew({ key: InputTypes.Select });
    wrapper.instance().handleAddNew({ key: InputTypes.NumberInput });
    wrapper.instance().handleAddNew({ key: InputTypes.TimePicker });
    wrapper.instance().handleAddNew({ key: InputTypes.DatePicker });
    wrapper.instance().handleAddNew({ key: InputTypes.Radio });
    wrapper.instance().handleAddNew({ key: InputTypes.Switch });
    expect(wrapper.state().config).toMatchSnapshot();
  });

  test('should update create form page config values', () => {
    const wrapper = shallow(
      <CreateForm
        form={mockForm}
        location={mockLocation}
        history={mockHistory}
      />
    );
    wrapper.instance().onChangeCurrentTag('NEW_TAG');
    expect(wrapper.state().currentTagInput).toEqual('NEW_TAG');
    wrapper.instance().resetCurrentTag();
    expect(wrapper.state().currentTagInput).toEqual('');
    wrapper.instance().changeConfig('NEW_CONFIG');
    expect(wrapper.state().config).toEqual('NEW_CONFIG');
  });

  test('should save new form', async (done) => {
    const wrapper = shallow(
      <CreateForm
        form={mockForm}
        location={{
          search: ''
        }}
        history={mockHistory}
      />
    );
    setTimeout(async () => {
      await wrapper.instance().onSubmit({ preventDefault: () => {} });
      setTimeout(() => {
        expect(message.success).toHaveBeenCalledWith('Saved Form Successfully!');
        expect(mockPush).toHaveBeenCalledWith(RoutePaths.MyForms);
        expect(API.post.mock.calls[0]).toMatchSnapshot(0);
        done();
      }, 1100);
    });
  });

  test('should update current form', async (done) => {
    const wrapper = shallow(
      <CreateForm
        form={mockForm}
        location={mockLocation}
        history={mockHistory}
      />
    );
    setTimeout(async () => {
      await wrapper.instance().onSubmit({ preventDefault: () => {} });
      setTimeout(() => {
        expect(message.success).toHaveBeenCalledWith('Saved Form Successfully!');
        expect(mockPush).toHaveBeenCalledWith(RoutePaths.MyForms);
        expect(API.put.mock.calls[0]).toMatchSnapshot(0);
        done();
      }, 1100);
    });
  });

});