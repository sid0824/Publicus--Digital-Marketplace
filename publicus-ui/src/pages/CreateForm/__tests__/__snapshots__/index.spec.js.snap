// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create Form should generate initial configs for all input types 1`] = `
Object {
  "1": Object {
    "prompt": "What is Your... ?",
    "rules": Array [],
    "type": "textInput",
  },
  "2": Object {
    "options": Array [],
    "prompt": "What is Your... ?",
    "rules": Array [],
    "type": "select",
  },
  "3": Object {
    "incrementor": 1,
    "max": 100,
    "min": 0,
    "prompt": "What is Your... ?",
    "rules": Array [],
    "startingNumber": 0,
    "type": "numberInput",
  },
  "4": Object {
    "prompt": "What time is the...?",
    "rules": Array [],
    "type": "timePicker",
  },
  "5": Object {
    "prompt": "What date is the...?",
    "rules": Array [],
    "showTime": false,
    "type": "datePicker",
  },
  "6": Object {
    "prompt": "What is Your... ?",
    "radioOptions": Array [
      "Placeholder",
    ],
    "rules": Array [],
    "type": "radio",
  },
  "7": Object {
    "defaultValue": false,
    "prompt": "What is Your... ?",
    "rules": Array [],
    "type": "switch",
  },
}
`;

exports[`Create Form should render the create form page 1`] = `
<Form
  colon={true}
  hideRequiredMark={false}
  layout="horizontal"
  onSubmit={[Function]}
  style={
    Object {
      "marginLeft": "calc(50% - 250px)",
      "marginRight": "calc(50% - 250px)",
      "marginTop": 10,
    }
  }
>
  <h1>
    Form Details
  </h1>
  Form Name
  <FormItem
    hasFeedback={false}
  >
    <Input
      placeholder="Enter the form name"
      style={
        Object {
          "width": 500,
        }
      }
      type="text"
    />
  </FormItem>
  Form Description
  <FormItem
    hasFeedback={false}
  >
    <TextArea
      placeholder="Enter the form description"
      rows={3}
      style={
        Object {
          "width": 500,
        }
      }
    />
  </FormItem>
  Form Price ($)
  <FormItem
    hasFeedback={false}
  >
    <InputNumber
      max={500}
      min={0}
      step={1}
      style={
        Object {
          "width": 500,
        }
      }
    />
  </FormItem>
  Form Tags
  <FormItem
    hasFeedback={false}
  >
    <Select
      choiceTransitionName="zoom"
      dropdownRender={[Function]}
      mode="multiple"
      onChange={[Function]}
      onSearch={[Function]}
      placeholder="Add tags to the form"
      showSearch={false}
      size="default"
      style={
        Object {
          "width": 500,
        }
      }
      transitionName="slide-up"
    />
  </FormItem>
  Redirect URL (Optional)
  <FormItem
    hasFeedback={false}
  >
    <Input
      placeholder="Enter a url to redirect on form complete"
      style={
        Object {
          "width": 500,
        }
      }
      type="text"
    />
  </FormItem>
  User Instructions (Optional)
  <FormItem
    hasFeedback={false}
  >
    <TextArea
      placeholder="Enter instructions for users at the top of the form"
      rows={3}
      style={
        Object {
          "width": 500,
        }
      }
    />
  </FormItem>
  <br />
  <Form(CreateFormFields)
    changeConfig={[Function]}
    config={
      Object {
        "1": Object {
          "prompt": "What is Your... ?",
          "rules": Array [
            "required",
          ],
          "type": "textInput",
        },
        "2": Object {
          "defaultValue": false,
          "prompt": "What is Your... ?",
          "rules": Array [
            "required",
          ],
          "type": "switch",
        },
      }
    }
  />
  <br />
  <br />
  <div
    style={
      Object {
        "display": "flex",
        "flexDirection": "row",
      }
    }
  >
    <FormItem
      hasFeedback={false}
    >
      <Button
        block={false}
        ghost={false}
        htmlType="submit"
        loading={false}
        style={
          Object {
            "marginRight": 5,
          }
        }
        type="primary"
      >
        Save
      </Button>
    </FormItem>
    <Dropdown
      mouseEnterDelay={0.15}
      mouseLeaveDelay={0.1}
      overlay={
        <Menu
          onClick={[Function]}
        >
          <MenuItem>
            Text Input
          </MenuItem>
          <MenuItem>
            Number Input
          </MenuItem>
          <MenuItem>
            Select
          </MenuItem>
          <MenuItem>
            Switch
          </MenuItem>
          <MenuItem>
            Time Picker
          </MenuItem>
          <MenuItem>
            Date Picker
          </MenuItem>
          <MenuItem>
            Radio Button
          </MenuItem>
        </Menu>
      }
      placement="bottomLeft"
      trigger={
        Array [
          "click",
        ]
      }
    >
      <Button
        block={false}
        ghost={false}
        htmlType="button"
        loading={false}
        style={
          Object {
            "marginTop": 4,
          }
        }
      >
        Create New Field 
        <Icon
          type="down"
        />
      </Button>
    </Dropdown>
  </div>
  <br />
  <br />
  <br />
</Form>
`;

exports[`Create Form should save new form 1`] = `
Array [
  "prod-publicus-app-api",
  "/survey",
  Object {
    "body": Object {
      "MOCK": "VALUE",
      "form": Object {},
    },
  },
]
`;

exports[`Create Form should update current form 1`] = `
Array [
  "prod-publicus-app-api",
  "/survey/4fe4b2c0-f4a5-11e9-87b0-6785951dc893",
  Object {
    "body": Object {
      "MOCK": "VALUE",
      "form": Object {
        "1": Object {
          "prompt": "What is Your... ?",
          "rules": Array [
            "required",
          ],
          "type": "textInput",
        },
        "2": Object {
          "defaultValue": false,
          "prompt": "What is Your... ?",
          "rules": Array [
            "required",
          ],
          "type": "switch",
        },
      },
      "surveyId": "4fe4b2c0-f4a5-11e9-87b0-6785951dc893",
    },
  },
]
`;
